### DISCUSSION TITLE
Variable size value types

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/runtime#94833] [Ideas] [answered]
Currently, at least from what I know, value types a fixed in size. Especially `structs` that can have arrays do not allow to have a dynamic size of an array, its needs to be fixed, either by using the `fixed` keyword or by the newly introduced `InlineArrays`. In many high-performance though, a variable size is required for example a `Vector` type. Where the current implementation in its many variants (Vector2, Vector3, Vector3D, etc) surely make sense, it is often required to have a dynamic variant of these. 

But very often you have an array of vectors, you need to be able to marshal somehow to structs. So for example we have images where each pixel can contain one component (grayscale), 3 components (RGB), 4 components (RGBA). Currently each needs to be treated differently. Additionally, the datatype can change, so simply writing a switch-case around it using either a `float`, `Vector3` or `Vector4` is not the solution, since they can only contain `floats`.

There are other requests that have similar requirements https://github.com/dotnet/csharplang/discussions/7508. But obviously the runtime needs to do some work. My proposal would be to allow something like

```cs
public readonly ref struct FixedStruct<T, size> where T : struct where size : const int
{
    public readonly Span<T> Value = stackalloc T[size]
}
```

But I could also live with this

```cs
public readonly ref struct FixedStruct<T> where T : struct
{
    public readonly Span<T> Value;
    public FixedStruct(int size)
    {
        Value = stackalloc byte[size];
    }
}
```

It is currently only a proposal which might not have anything to do with reality. My idea was that the stackalloc inside the struct gets part of the structs stacksize, but there might be better solutions.

**Comment by Sergio0694** [comment, marked as answer]
Duplicate of https://github.com/dotnet/runtime/issues/89730?

**Reply by msedi** [reply]
Thanks @Sergio0694. This seems to cover my questions.
